src/libraries/System.Net.WebSockets/src/System/Net/WebSockets/WebSocketStream.cs

```diff
@@ -0,0 +1,379 @@
+            public override ValueTask WriteAsync(ReadOnlyMemory<byte> buffer, CancellationToken cancellationToken = default)
+            {
+                if (_disposed)
+                {
+                    return ValueTask.FromException(new ObjectDisposedException(GetType().FullName));
+                }
+
+                if (!CanWrite)
+                {
+                    return ValueTask.FromException(new NotSupportedException(SR.NotWriteableStream));
+                }
+
+                if (cancellationToken.IsCancellationRequested)
+                {
+                    return ValueTask.FromCanceled(cancellationToken);
+                }
+
+                return WebSocket.SendAsync(buffer, _messageType, endOfMessage: false, cancellationToken);
+            }
+
+            public override ValueTask DisposeAsync()
+            {
+                if (!_disposed)
+                {
+                    _disposed = true;
+                    return WebSocket.SendAsync(ReadOnlyMemory<byte>.Empty, _messageType, endOfMessage: true, CancellationToken.None);
+                }
+
+                return default;
+            }
+        }
+
+        /// <summary>Provides a stream that wraps a <see cref="WebSocket"/> and reads a single message.</summary>
+        private sealed class ReadMessageStream(WebSocket webSocket) : WebSocketStream(webSocket)
+        {
+            /// <summary>Whether we've seen and end-of-message marker.</summary>
+            private bool _eof;
+
+            /// <inheritdoc />
+            public override bool CanWrite => false;
+
+            /// <inheritdoc />
+            public override async ValueTask<int> ReadAsync(Memory<byte> buffer, CancellationToken cancellationToken = default)
+            {
+                ObjectDisposedException.ThrowIf(_disposed, this);
+
+                if (!CanRead)
+                {
+                    throw new NotSupportedException(SR.NotReadableStream);
+                }
+
+                cancellationToken.ThrowIfCancellationRequested();
+
+                while (!_eof && WebSocket.State < WebSocketState.CloseReceived)
+                {
+                    ValueWebSocketReceiveResult result = await WebSocket.ReceiveAsync(buffer, cancellationToken).ConfigureAwait(false);
+                    if (result.MessageType is WebSocketMessageType.Close)
+                    {
+                        break;
+                    }
+
+                    if (result.EndOfMessage)
+                    {
+                        _eof = true;
+                    }
+
+                    if (result.Count > 0 || buffer.IsEmpty)
+                    {
+                        return result.Count;
+                    }
+                }
+
+                return 0;
+            }
+
+            /// <inheritdoc />
+            public override ValueTask DisposeAsync()
+            {
+                _disposed = true;
+                return default;
+            }
```

Comment:
If I [understood Natalia correctly](https://github.com/dotnet/runtime/issues/111217#:~:text=Abort%20the%20connection%20immediately%20if%20a%20message%20is%20disposed%20before%20full%20consumption%2C%20to%20avoid%20corruption%20and%20support%20multiplexing.%20This%20aligns%20with%20WebSocket%20cancellation%20semantics.), we might want to abort the WebSocket here if we haven't hit EOF.